File: tictactoe_back/src/main/java/com/mkierzkowski/tictactoe_back/service/game/GameService.java
package com.mkierzkowski.tictactoe_back.service.game;

import com.mkierzkowski.tictactoe_back.dto.request.GameBoardMoveRequestDto;
import com.mkierzkowski.tictactoe_back.dto.response.ranks.UserRankResponseDto;
import com.mkierzkowski.tictactoe_back.model.game.Game;
import org.springframework.validation.Errors;

import java.util.List;

public interface GameService {
    List<Game> getAvailableGames();
    Game createNewGame();
    Game getGameById(Long gameId);
    void joinGame(Long gameId);
    void leaveGame(Long gameId);
    void boardMove(Long gameId, GameBoardMoveRequestDto gameBoardMoveRequestDto, Errors errors);
    List<UserRankResponseDto> getRanks();
}

File: tictactoe-front/src/app/game/services/game.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from "@angular/common/http";
import { AvailableGame } from "../models/AvailableGame";
import { environment } from "../../../environments/environment";
import { map } from "rxjs/operators";
import { Observable } from "rxjs";
import { CreateNewGameResponse } from "../models/CreateNewGameResponse";
import { GameDetails } from "../models/GameDetails";
import { BoardSquare } from "../models/BoardSquare";
import { UserRank } from "../models/UserRank";

@Injectable({
  providedIn: 'root'
})
export class GameService {

  readonly apiURL = `${environment.apiUrl}/game`;

  constructor(
    private http: HttpClient,
  ) { }

  getAvailableGames() : Observable<AvailableGame[]> {
    return this.http.get<any>(`${this.apiURL}/available`, {withCredentials: true})
      .pipe(map(response => response.availableGames));
  }

  getGameDetails(gameId: string) : Observable<GameDetails> {
    return this.http.get<any>(`${this.apiURL}/${gameId}`,{withCredentials: true});
  }

  performMove(gameId: string, boardSquare: BoardSquare) : Observable<any> {
    return this.http.post<any>(`${this.apiURL}/${gameId}/boardMove`,{boardSquareId: boardSquare.squareId},{withCredentials: true});
  }

  createNewGame() : Observable<CreateNewGameResponse> {
    return this.http.post<CreateNewGameResponse>(`${this.apiURL}/new`, {},{withCredentials: true});
  }

  joinGame(gameId: string) : Observable<any> {
    return this.http.post<any>(`${this.apiURL}/${gameId}/join`,{},{withCredentials: true});
  }

  leaveGame(gameId: string) : Observable<any> {
    return this.http.post<any>(`${this.apiURL}/${gameId}/leave`,{},{withCredentials: true});
  }

  getUsersRanks(): Observable<UserRank[]> {
    return this.http.get<any>(`${this.apiURL}/ranks`, {withCredentials: true})
      .pipe(map(response => response.usersRanks));
  }
}
